### DISCUSSION TITLE
How to improve Interaction to Next Paint (INP)

### DISCUSSION THREAD

**Discussion by simonferndriger** [alpinejs/alpine#3954] [1. Help]
Currently, our web app (using AlpineJS) has an INP value of ~230ms (on mobile devices). Everything above 200ms should be improved. Are there any best practices that have a positive impact on INP regarding AlpineJS implementation?

**Comment by ekwoka** [comment]
This is very broad as not all interactions are the same.

The only things I could imagine Alpine missing a frame on would involve particularly large x-for being changed.

If you access a getter that has a lot of work quite often but it's not changing, finding a way to memoize it can be worthwhile

**Reply by simonferndriger** [reply]
**Large x-for**
Yes I do have a large x-for (and I load more into it with infinity scroll). And if a category is selected, the large x-for array gets replaced completely. Can this perhaps be optimized?

**Accessing Getter**
Does this only concern real getters or functions in general? On `init()` I for example use `this.w = () => this.webcam || {};` to avoid js errors because of the template issue not cleaning up in the right order. Would it perform better to just use `this.webcam?.*` or to reset `this.w` with every `this.webcam` change using `$watch('webcam')`?

Happy New Year 😀

**Reply by simonferndriger** [reply]
However, we also have video ads within this (large) x-for list. That causes a bit of a jerk while scrolling (I guess ads will always be a performance problem), maybe that is also counting as INP...

**Reply by ekwoka** [reply]
the webacom thing `this.webcam?.prop` is best, but it's minor. That's not meaninful work regardless.

The large x-for would be the main thing for sure.

**Reply by simonferndriger** [reply]
Thanks. Any advice for the large x-for? Or is the only advice, don't do it? 😉

PS: I removed the ads within the large x-for from the dom now as soon as they are out of the viewport, this already seems to help with scrolling performance.

**Reply by ekwoka** [reply]
It depends on the size of the x-for.

Mostly the cost of the process of updating an x-for is primarily in the creation of entries. So you can make it add new items in smaller groups, so that they won't create nearly as many at a time.

**Reply by simonferndriger** [reply]
Alright, thanks! So I only add 50 items to the array each time - that's not too much is it?

And, does it matter how I extend the array for the x-for? Currently I'm using `webcams = [...webcams, ...data.webcams]`.

**Reply by ekwoka** [reply]
Pushing is better, but it mostly doesn't matter provided they have keys.

50 should be fine (but the best amount depends also on what all is done for the template for each item)

**Reply by simonferndriger** [reply]
Thanks! By _keys_ you mean for the `x-for` I assume? I.e. ` <template :key="w.webcamId" x-for="(w, index) in webcams" x-bind="$" x-data="rasterAd">`

Alright, thanks a lot! Let's see how Google Search Console is reacting to these changes in the next weeks :)

Happy New Year
Simon
